### DISCUSSION TITLE
how to terminate parsing of an error

### DISCUSSION THREAD

**Discussion by dfgordon** [tree-sitter/tree-sitter#1556] [Q&A]
I'm finding Tree-sitter consumes good code as part of an error.  In my case I would like it to tell it to give up trying to resolve the error when it hits a new line.  Here is a specific example.

First without any error:

```bas
10 X = SIN(A)
20 PRINT "HELLO"
```
this resolves the code into two lines just as desired:

```sexpr
(source_file [0, 0] - [2, 0]
  (line [0, 0] - [1, 0]
    (linenum [0, 0] - [0, 3])
    (statement [0, 3] - [0, 13]
      (assignment [0, 3] - [0, 13]
        (realvar [0, 3] - [0, 5])
        (eq_tok [0, 5] - [0, 6])
        (fcall [0, 6] - [0, 13]
          (sin_tok [0, 6] - [0, 10])
          (realvar [0, 11] - [0, 12])))))
  (line [1, 0] - [2, 0]
    (linenum [1, 0] - [1, 3])
    (statement [1, 3] - [1, 16]
      (print_tok [1, 3] - [1, 8])
      (string [1, 8] - [1, 16]))))
```

Now introduce an error in the first line (error is we can't take sine of a string):

```bas
10 X = SIN(A$)
20 PRINT "HELLO"
```
In this case the parser blows by the newline, finding no valid code at all, even though the second line is valid:

```sexpr
(source_file [0, 0] - [2, 0]
  (ERROR [0, 0] - [2, 0]
    (linenum [0, 0] - [0, 3])
    (realvar [0, 3] - [0, 5])
    (eq_tok [0, 5] - [0, 6])
    (sin_tok [0, 6] - [0, 10])
    (ERROR [0, 11] - [1, 8]
      (svar [0, 11] - [0, 13])
      (linenum [1, 0] - [1, 3])
      (print_tok [1, 3] - [1, 8]))
    (string [1, 8] - [1, 16])))
../vscode-language-applesoft/sample/test.bas	0 ms	(ERROR [0, 0] - [2, 0])
```

The `line` rule is of the form `seq($._valid_stuff,$._newline)`, cf. #1424.

A simple workaround would be to call the parser on one line at a time from within this or that application, but I was wondering if there is a way to get TS to do this internally, or better, if there is a general way to handle this class of problem.

**Comment by ahelwer** [comment]
See https://github.com/tree-sitter/tree-sitter/discussions/1205
tl;dr can't be done

**Comment by dfgordon** [comment]
Thanks, based on the #1205 chain it appears the answer is "you can't today, but better error recovery is possible in principle and people are thinking about it."
